libpod/events/logfile.go

```diff
@@ -107,3 +115,125 @@ func (e EventLogFile) Read(ctx context.Context, options ReadOptions) error {
 func (e EventLogFile) String() string {
 	return LogFile.String()
 }
+
+// Rotates the log file if the log file size and content exceeds limit
+func rotateLog(logfile string, content string, limit uint64) (bool, error) {
+	e := &Event{}
+	if limit == 0 {
+		return false, nil
+	}
+	file, err := os.Stat(logfile)
+	if err != nil {
+		return false, err
+	}
+	var filesize = uint64(file.Size())
+	var contentsize = uint64(len([]rune(content)))
+	if filesize+contentsize >= limit {
+		if err := truncate(logfile); err != nil {
+			return false, err
+		}
+		rotateEvent := e.RotateEvent(Rotate)
+		rotateEventJSONString, err := json.Marshal(rotateEvent)
+		if err != nil {
+			return false, nil
+		}
+		rotateEventString := string(rotateEventJSONString)
+		f, err := os.OpenFile(logfile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0700)
+		if err != nil {
+			return false, err
+		}
+		if _, err := f.WriteString(fmt.Sprintf("%s\n", rotateEventString)); err != nil {
+			return false, err
+		}
+		return true, nil
+	}
+	return false, nil
+}
+
+// Truncates the log file and saves 50% of content to new log file
+func truncate(path string) error {
+	orig, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+
+	origFinfo, err := orig.Stat()
+	if err != nil {
+		return err
+	}
+
+	size := origFinfo.Size()
+	threshold := uint64(size / 2)
+
+	tmp, err := ioutil.TempFile("", "")
+	if err != nil {
+		return err
+	}
+	var readBytes uint64
+	scanner := bufio.NewScanner(orig)
+	for scanner.Scan() {
+		rawLine := scanner.Bytes()
+		rawLine = append(rawLine, '\n')
+		readBytes += uint64(len(rawLine))
+		if readBytes <= threshold {
+			// Skip the line if we're still below 50% of the file.
+			continue
+		}
+		if _, err := tmp.Write(rawLine); err != nil {
+			return fmt.Errorf("writing to temporary file: %w", err)
+		}
+	}
+	if err := orig.Close(); err != nil {
+		return err
+	}
+	if err := tmp.Close(); err != nil {
+		return err
+	}
+	return renameLog(tmp.Name(), path)
+}
+
+// Renames from, to
+func renameLog(from, to string) error {
+	err := os.Rename(from, to)
+	if err == nil {
+		return nil
+	}
+
+	if !strings.Contains(err.Error(), "invalid cross-device link") {
+		return err
+	}
+
+	// Files are not on the same partition, so we need to copy them the
```

Comment:
Why do we use this workaround? We should just write the tmp event file right next to the event log. In this case we should never run into rename should not fail.

Reply:
Can you elaborate on what you mean by "right next"?  How do we know we can read/write to that location?

Reply:
We need to write the tmp file in the same directory as the event log file. In this case os.Rename should always work and we do not have to worry about copying the data another time.

Reply:
see the logic in `c/strorage/pkg/ioutils` I guess we could use it here as well: https://github.com/containers/storage/blob/5d55c443125e7383cbc125afdfd257d332c0e507/pkg/ioutils/fswriters.go#L30

Reply:
> We need to write the tmp file in the same directory as the event log file. In this case os.Rename should always work and we do not have to worry about copying the data another time.

That is rather unexpected behavior (to me).  I don't want a tool to write to user directories.

Reply:
Why is this unexpected? You favourite editor vim does this as well with the .swp files for example.
This is done in may tools to make sure that you do not have to copy data because rename will fail if the paths are on other file systems.

Reply:
I am still not convinced that it is a good idea to write to a temporary file at a location on a different file system. By doing this we have a) double the writes/reads than necessary and b) the rename is no longer atomic, thus a reader could end up with partially written file.

Reply:
> You favourite editor vim does this as well with the .swp files for example.

In case of vim it's even worse since it may create files in a git repository.  I never liked the behavior.  There's no guarantee that we could create a file in a custom user location.

> I am still not convinced that it is a good idea to write to a temporary file at a location on a different file system. By doing this we have a) double the writes/reads than necessary 

I do not worry that much about the additional OPs.  It's not happening often.

> and b) the rename is no longer atomic, thus a reader could end up with partially written file.

That is a good point.

> https://github.com/containers/storage/blob/5d55c443125e7383cbc125afdfd257d332c0e507/pkg/ioutils/fswriters.go#L30

The issue with _this_ atomic writer is that all content must be read to memory, it's not buffered, and events file may be arbitrary large.

Reply:
We do not have to have to use ioutlis. It should be enough to write the tmp file in the same directory as the event log, in this case the simple os.Rename will work and we do not need the fallback code.

Reply:
> We do not have to have to use ioutlis. It should be enough to write the tmp file in the same directory as the event log, in this case the simple os.Rename will work and we do not need the fallback code.

There is no guarantee we have permissions to create a new file in the same directory.  The user may specify a custom location where it can read/write the events file but has no permission in the directory.

Reply:
That could happen but I would argue it is unlikely. In such case we could still fall back to the current logic.

I believe that the risk of a partial read is much bigger with the current version.

Reply:
Sounds like a good compromise :+1: @nicrowe00, you can use the diff below (it'll also fix some indent issues):
```diff
diff --git a/libpod/events/logfile.go b/libpod/events/logfile.go
index 39eccb64a1d9..5091f3723a13 100644
--- a/libpod/events/logfile.go
+++ b/libpod/events/logfile.go
@@ -10,11 +10,12 @@ import (
 	"io"
 	"io/ioutil"
 	"os"
+	"path"
 	"time"
 
 	"github.com/containers/podman/v4/pkg/util"
 	"github.com/containers/storage/pkg/lockfile"
-       "github.com/nxadm/tail"
+	"github.com/nxadm/tail"
 	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
 	"golang.org/x/sys/unix"
@@ -74,14 +75,14 @@ func (e EventLogFile) writeString(s string) error {
 }
 
 func (e EventLogFile) getTail(options ReadOptions) (*tail.Tail, error) {
-       reopen := true
-       seek := tail.SeekInfo{Offset: 0, Whence: os.SEEK_END}
-       if options.FromStart || !options.Stream {
-               seek.Whence = 0
-               reopen = false
-       }
-       stream := options.Stream
-       return tail.TailFile(e.options.LogFilePath, tail.Config{ReOpen: reopen, Follow: stream, Location: &seek, Logger: tail.DiscardingLogger, Poll: true})
+	reopen := true
+	seek := tail.SeekInfo{Offset: 0, Whence: os.SEEK_END}
+	if options.FromStart || !options.Stream {
+		seek.Whence = 0
+		reopen = false
+	}
+	stream := options.Stream
+	return tail.TailFile(e.options.LogFilePath, tail.Config{ReOpen: reopen, Follow: stream, Location: &seek, Logger: tail.DiscardingLogger, Poll: true})
 }
 
 // Reads from the log file
@@ -172,8 +173,8 @@ func rotateLog(logfile string, content string, limit uint64) (bool, error) {
 }
 
 // Truncates the log file and saves 50% of content to new log file
-func truncate(path string) error {
-	orig, err := os.Open(path)
+func truncate(filePath string) error {
+	orig, err := os.Open(filePath)
 	if err != nil {
 		return err
 	}
@@ -187,9 +188,14 @@ func truncate(path string) error {
 	size := origFinfo.Size()
 	threshold := size / 2
 
-	tmp, err := ioutil.TempFile("", "")
+	tmp, err := ioutil.TempFile(path.Dir(filePath), "")
 	if err != nil {
-		return err
+		// Retry in /tmp in case creating a tmp file in the same
+		// directory has failed.
+		tmp, err = ioutil.TempFile("", "")
+		if err != nil {
+			return err
+		}
 	}
 	defer tmp.Close()
 
@@ -207,8 +213,8 @@ func truncate(path string) error {
 		return fmt.Errorf("writing truncated contents: %w", err)
 	}
 
-	if err := renameLog(tmp.Name(), path); err != nil {
-		return fmt.Errorf("writing back %s to %s: %w", tmp.Name(), path, err)
+	if err := renameLog(tmp.Name(), filePath); err != nil {
+		return fmt.Errorf("writing back %s to %s: %w", tmp.Name(), filePath, err)
 	}
 
 	return nil
```